Plutarch/Lift.hs

```diff
@@ -0,0 +1,77 @@
+{-# LANGUAGE UndecidableInstances #-}
+
+module Plutarch.Lift (
+  PConstant (..),
+  PLift (..),
+  plift,
+  LiftError (..),
+  PDefaultUniType,
+) where
+
+import Data.Bifunctor (first)
+import Data.Data (Proxy (Proxy))
+import Data.Text
+import qualified Data.Text as T
+import GHC.Stack (HasCallStack)
+import Plutarch
+import Plutarch.Evaluate (evaluateScript)
+import Plutarch.Internal (punsafeConstant)
+import Plutarch.Prelude
+import qualified Plutus.V1.Ledger.Scripts as Scripts
+import qualified PlutusCore as PLC
+import PlutusCore.Constant (readKnownSelf)
+import qualified PlutusCore.Constant as PLC
+import PlutusCore.Evaluation.Machine.Exception (
+  EvaluationException,
+  MachineError,
+ )
+import qualified UntypedPlutusCore as UPLC
+import UntypedPlutusCore.Evaluation.Machine.Cek (CekUserError)
+
+class PConstant p (h :: Type) where
+  -- {-
+  -- Create a Plutarch-level constant, from a Haskell value.
+  --
+  -- Example:
+  -- > pconstant @PInteger 42
+  -- -}
+  pconstant :: h -> Term s p
+
+data LiftError
+  = LiftError_ScriptError Scripts.ScriptError
+  | LiftError_EvalException T.Text -- Using Text, because there is no Eq possible with DeBruijn naming.
+  deriving stock (Eq, Show)
+
+class PLift (h :: Type) p where
+  -- {-
+  -- Convert a Plutarch term to the associated Haskell value. Fail otherwise.
+  -- This will fully evaluate the arbitrary closed expression, and convert the
+  -- resulting value.
+  -- -}
+  plift' :: ClosedTerm p -> Either LiftError h
+
+plift :: (PLift h p, HasCallStack) => ClosedTerm p -> h
+plift prog = either (error . show) id $ plift' prog
+
+instance (PLC.DefaultUni `PLC.Contains` h, PDefaultUniType p ~ h) => PConstant p h where
+  pconstant =
+    punsafeConstant . PLC.Some . PLC.ValueOf (PLC.knownUniOf (Proxy @h))
+
+instance PLC.KnownTypeIn PLC.DefaultUni (UPLC.Term PLC.DeBruijn PLC.DefaultUni PLC.DefaultFun ()) h => PLift h p where
+  plift' prog =
+    case evaluateScript (compile prog) of
```

Comment:
Can we somehow move this logic into `plift` and make `plift'` as small as possible?

Reply:
Not sure how. It is essentially a wrapper around `evaluateScript` and `readKnownSelf` (with some error wrapping) which we need in the instance. And the type is already small enough I think:

```haskell
plift' :: ClosedTerm p -> Either LiftError h
```

Reply:
Taking a step back: do we even really need these two type classes (`PConstant` and `PLift`)? They are never instantiated outside the module.

Reply:
For one, I think you can likely merge them.

Reply:
I think we might also want *not* to export them. You could export them from an internal module. Downstream users should not define their own instances.

Reply:
Also, you might actually be able to remove the type classes and just use the same constraints as used for `readKnownSelf` and such.

Reply:
Done. The type class is no more.